src/main/java/org/jabref/gui/maintable/MainTableDataModel.java

```diff
@@ -16,73 +14,132 @@
+        this.fieldValueFormatter = new SimpleObjectProperty<>(new MainTableFieldValueFormatter(nameDisplayPreferences, bibDatabaseContext));
 
         resetFieldFormatter();
 
         ObservableList<BibEntry> allEntries = BindingsHelper.forUI(context.getDatabase().getEntries());
         ObservableList<BibEntryTableViewModel> entriesViewModel = EasyBind.mapBacked(allEntries, entry ->
-                new BibEntryTableViewModel(entry, bibDatabaseContext, fieldValueFormatter));
+                new BibEntryTableViewModel(entry, bibDatabaseContext, fieldValueFormatter, stateManager));
 
         entriesFiltered = new FilteredList<>(entriesViewModel);
         entriesFiltered.predicateProperty().bind(
-                EasyBind.combine(stateManager.activeGroupProperty(),
+                EasyBind.combine(
+                        stateManager.activeGroupProperty(),
                         stateManager.activeSearchQueryProperty(),
                         groupsPreferences.groupViewModeProperty(),
-                        (groups, query, groupViewMode) -> entry -> isMatched(groups, query, entry))
+                        (groups, query, groupViewMode) -> {
+                            // TODO btut: do not repeat search if display mode changes. Check if the same can be done for groups
+                            doSearch(query);
+                            updateSearchGroups(stateManager, bibDatabaseContext, luceneManager);
+                            return entry -> isMatched(groups, query, entry);
+                        })
         );
 
-        IntegerProperty resultSize = new SimpleIntegerProperty();
-        resultSize.bind(Bindings.size(entriesFiltered));
-        stateManager.setActiveSearchResultSize(context, resultSize);
+        entriesViewModel.addListener((ListChangeListener<BibEntryTableViewModel>) c -> {
+            if (stateManager.activeSearchQueryProperty().isPresent().get()) {
+                while (c.next()) {
+                    if (c.wasAdded()) {
+                        doSearch(stateManager.activeSearchQueryProperty().get());
+                        return;
+                    }
+                }
+            }
+        });
+
         // We need to wrap the list since otherwise sorting in the table does not work
         entriesFilteredAndSorted = new SortedList<>(entriesFiltered);
     }
 
+    public void removeBindings() {
+        entriesFiltered.predicateProperty().unbind();
+    }
+
+    public static void updateSearchGroups(StateManager stateManager, BibDatabaseContext bibDatabaseContext, LuceneManager luceneManager) {
+        stateManager.getSelectedGroups(bibDatabaseContext)
+                    .stream()
+                    .map(GroupTreeNode::getGroup)
+                    .filter(SearchGroup.class::isInstance)
+                    .map(SearchGroup.class::cast)
+                    .forEach(group -> group.updateMatches(luceneManager));
+    }
+
+    private void doSearch(Optional<SearchQuery> query) {
```

Comment:
Please try to avoid to pass `Optional`.

Maybe its ok here, because of the `lastSearchQuery`. But then please document it in the JavaDoc that it is necessary.

In general, it should be avoided. Normal Java style: create two methods: One with the parameter and another one without.